### ISSUE TITLE
Implement Access Control List for job folders

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#448]
## Requirements

Every time a workflow is applied, a folder is created for the corresponding job (a typical name would look like `workflow_0007_job_0002`). 

In a single-user ("user" as in "user of the machine") use case, there is nothing special about this folder. In a SLURM multi-user case, there are two different users ("users" as in "users of the cluster") that need to have read/write access to this folder, as part of this scheme:
1. `fractal`  (or whatever user is running the server) creates the folder
2. `fractal` writes some logs.
3. `fractal` writes some files that are necessary for running the task (e.g. the sbatch script and a serialized version of the task to run).
5. `test01` (or whatever user needs to run the task) reads the files from point 2, and runs the task.
6. `test01` writes some logs, and some files with the task output.
7. `fractal` reads the output files and proceeds.
8. `fractal` need to be able to access any other log files (created by `test01`) in that folder.

## Current status
At the moment, `fractal` provides broad (777) access to the job folder, and `test01` provides broad (777) access to their logs. This works, but it's obviously not the way to go (ref #317)

## Proposed change

We add a new config variable (provisional name `FRACTAL_ACL_OPTIONS` defaulting to `none`), and proceed differently based on its value:
* If `FRACTAL_ACL_OPTIONS="none"`, folders and files are created with standard permissions (i.e. we never set them to 777). This is meant to only work for a single-user use case, where the user has access to all of their folders/file.
* If `FRACTAL_ACL_OPTIONS="standard"` (or a better name..), we assume that the `setfacl` command is available on the machine running the server (we can also check it by hand, to be sure), and we proceed as in https://github.com/fractal-analytics-platform/fractal-server/issues/317#issuecomment-1352836505. That is, we use standard ACLs to provide access to the job folder to **two users only** (`fractal` and `test01`). This is done upon folder creation, and then we don't touch permissions any more (neither on the `fractal` side, nor on the `test01` side).
* If `FRACTAL_ACL_OPTIONS="nfs4"`, we follow the same logic but with nfs4-specific commands (as in #447).

**Comment by jluethi** [comment]
Sounds like a good plan to me, thanks for writing it up @tcompa 

@enricotagliavini When you're better again, could you let us know whether such a setup will pass the FMI requirements? At FMI, we'd be running Fractal in `FRACTAL_ACL_OPTIONS="standard"` mode, thus using the ACLs for access control (see option A in this comment: https://github.com/fractal-analytics-platform/fractal-server/issues/317#issuecomment-1362519770)

**Comment by enricotagliavini** [comment]
Hello Joel and Fractal team,

as I mentioned in one of my last emails to you, the more I think about the ACL / permission approach the more I think it's walking into a rat's nest. Permission and ACLs are complex to understand and it's easy to make mistakes when changing them. They are also persistent, meaning that if you crash, or interrupt in the middle of the execution for whatever means, you end up in an undefined state, there is no automatic cleanup.

The best approach is with IPC / streaming, that's how modern programs (running as different users) exchange information when needed. The most common method I know about is by far D-Bus (which they are even try to include in the Linux Kernel itself for how much common this is... this of course caused quite a storm in the kernel community). It's safer, it's easier and it's more portable (or easier to make portable). Of course it's not the only option, you could simply stream the logs using SLURM itself if you prefer. By using some IPC / streaming technology you need to change no permission whatsoever, making no assumption or putting any requirement on the storage solution to be used.

As for your question I cannot answer without seeing the implementation first and having it tested on the field, trying to minimize the chances of overlooking something. One thing I can say is that the example in https://github.com/fractal-analytics-platform/fractal-server/issues/317#issuecomment-1352836505 seems not to set the default ACLs on the folder, so newly created files inside the folder will **not** have the same ACL as the parent, and might trigger permission issues later on. Another issue is that, at least in that comment, existing ACLs are not taken into consideration (e.g. other users might still have access by a group ACL). Given the task is just adding two users ACLs this is probably fine in most cases, but it probably needs careful review for each setup as it might have some unintended adverse effect. Another might be to reset all ACLs on that very folder before applying (```setfacl -b``` option), but that also might have unintended adverse effects, but it can be more strict if you reset, set all ACLs to "---" (000) except for the two users. As I mentioned a rat's nest.

Should you decide to proceed with ACLs I have a couple of advises to give:

- don't call it "standard", there are several ACL "standards". Call it "posix", which is the very most common in the Linux / Unix world and that's also what we use at FMI. The other two being "windows" and "nfs4". The latter being supported by pretty much no tool whatsoever (e.g. rsync and all GNU tools are all lacking support for "nfs4" style ACLs).
- instead of using the setfacl / getfacl command you might want to consider using the actual python wrapper library for libacl, which is the library setfacl / getfacl use: https://pylibacl.k1024.org/ . It's a fairly common python package. Creating a subprocess for setting the ACLs is a fairly rare operation, so it's ok if you prefer that, but it might be a bit more complicated to handle errors.

Thank you.
Kind regards.

**Comment by tcompa** [comment]
Thanks @enricotagliavini 

TL;DR
1. We (me and @mfranzon) have some questions on the IPC solution, and more generally we would rather stick with simpler solutions for now and consider IPC at later stage.
2. There seems to be a way to make the ACL solution work, requiring only 711 access to fractal (internal) folders. This looks like a decent compromise to us, since
    * 711 is already more restrictive than what many folders on clusters typically have;
    * The 711 folders only contain internal Fractal files, and not any (raw or processed) scientific data.
    * Any subfolder that needs broader permissions (i.e. because it should be accessible to both fractal and another user) is handled with ACL.
3. Thanks for other suggestions on the `standard->posix` renaming and on pylibacl. As for the latter: when I first looked at that library, I had the impression that it is not maintained very frequently (last release in 2020, very little activity on issues), but I'll have a second look.

---

Re: IPC solution

This is clearly an interesting option to explore, and we started looking at it with standard python tools (e.g. [ZeroMQ](https://zeromq.org)) which are also used by Parsl ([a project somewhat similar to Fractal](https://github.com/Parsl/parsl)). But we are a bit worried of starting with it right now, as it would add quite more complexity and slow down ongoing Fractal development. I would rather consider it as a more long term solution, that we could possibly prioritize a bit later down the road.

However, since we are already discussing this option, a couple of questions came up (these could very well come from my/our ignorance, or from the differences between zmq and d-bus):
1. Socket files are typically written/read by both users (fractal and the actual user), so that there must exist a folder where both users have access. Does this lead to similar issues as the one we are discussing right now (i.e. having a folder with very broad access)?
2. On a higher level, do you happen to know a project we could look at for an implementation of IPCs by different users? The closer to Fractal (e.g. python, multi-user,  HPC..) the better.
3. Looking at the way Parsl uses zmq, at some point they realized that these channels need to be secured in some way (encrypted and/or authenticated -- see e.g. [here](https://github.com/Parsl/parsl/issues/952) or [here](https://github.com/Parsl/parsl/issues/2199)). Do you have advice on best practices to proceed in this direction?

(note to self, as a future reference: let's also keep in mind RPC options like https://rpyc.readthedocs.io)

---

Re: ACL solution

The following example (produced via [this script](https://github.com/fractal-analytics-platform/fractal-server/blob/cf179d3e66f5b92ec780a019b7c2c4041c58f5dc/tests/sandbox_acl/example_acl.sh)) should be closer to your suggestions (although we can probably clean it up a bit more, e.g. by removing the two similar `setfacl` calls):

1. Set up folders
```console
fractal@exact:~$ mkdir /tmp/fractal
fractal@exact:~$ chmod 711 /tmp/fractal
fractal@exact:~$ mkdir /tmp/fractal/artifacts
fractal@exact:~$ chmod 711 /tmp/fractal/artifacts
fractal@exact:~$ mkdir /tmp/fractal/artifacts/workflow_0001_job_0001
fractal@exact:~$ chmod 700 /tmp/fractal/artifacts/workflow_0001_job_0001
```

2. Set ACL for `/tmp/fractal/artifacts/workflow_0001_job_0001`
```console
fractal@exact:~$ setfacl -b /tmp/fractal/artifacts/workflow_0001_job_0001
fractal@exact:~$ setfacl -k /tmp/fractal/artifacts/workflow_0001_job_0001
fractal@exact:~$ setfacl --recursive --modify user:fractal:rwx,user:test01:rwx,group::---,other::--- /tmp/fractal/artifacts/workflow_0001_job_0001
fractal@exact:~$ setfacl --default --recursive --modify user:fractal:rwx,user:test01:rwx,group::---,other::--- /tmp/fractal/artifacts/workflow_0001_job_0001
fractal@exact:~$ getfacl -p /tmp/fractal/artifacts/workflow_0001_job_0001
# file: /tmp/fractal/artifacts/workflow_0001_job_0001
# owner: fractal
# group: fractal
user::rwx
user:fractal:rwx
user:test01:rwx
group::---
mask::rwx
other::---
default:user::rwx
default:user:fractal:rwx
default:user:test01:rwx
default:group::---
default:mask::rwx
default:other::---
```

3. Verify that expected use cases (1. Write a file as `fractal`, read it as `test01`; 2. Write a file as `test01`, read it as `fractal`) do work:
```console
fractal@exact:~$ echo This-file-was-written-by-fractal > /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-fractal.txt
test01@exact:~$ cat /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-fractal.txt
This-file-was-written-by-fractal
fractal@exact:~$ getfacl -p /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-fractal.txt
# file: /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-fractal.txt
# owner: fractal
# group: fractal
user::rw-
user:fractal:rwx		#effective:rw-
user:test01:rwx			#effective:rw-
group::---
mask::rw-
other::---


test01@exact:~$ echo This-file-was-written-by-test01 > /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-test01.txt
fractal@exact:~$ cat /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-test01.txt
This-file-was-written-by-test01
test01@exact:~$ getfacl -p /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-test01.txt
# file: /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-test01.txt
# owner: test01
# group: test01
user::rw-
user:fractal:rwx		#effective:rw-
user:test01:rwx			#effective:rw-
group::---
mask::rw-
other::---
```

5. Verify that unexpected use cases (writing to the `artifacts` folder as a non-owner, or doing anything as a third user `test02`) are not authorized
```console
test01@exact:~$ touch /tmp/fractal/artifacts/some-file-2.txt
touch: cannot touch '/tmp/fractal/artifacts/some-file-2.txt': Permission denied

test02@exact:~$ touch /tmp/fractal/artifacts/some-file-3.txt
touch: cannot touch '/tmp/fractal/artifacts/some-file-3.txt': Permission denied

test02@exact:~$ touch /tmp/fractal/artifacts/workflow_0001_job_0001/some-file.txt
touch: cannot touch '/tmp/fractal/artifacts/workflow_0001_job_0001/some-file.txt': Permission denied

test02@exact:~$ cat /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-fractal.txt
cat: /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-fractal.txt: Permission denied

test02@exact:~$ cat /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-test01.txt
cat: /tmp/fractal/artifacts/workflow_0001_job_0001/file-of-test01.txt: Permission denied
```

Does this sound reasonable?

We are then preparing an implementation of this as part of Fractal, where the main part (the mkdr/chmod/setfacl) is done in a python function, using python function as much as possible (for sure for mkdir and chmod, and possibly also for the ACL part via pylibacl).

**Comment by enricotagliavini** [comment]
What about SLURM streaming at the beginning rather than ACLs, as a first simple implementation? Fractal can run any job as the user, so you can run:

```srun -n 1 -c 1 --mem 1G --time '5:0' -dependency afterany:<sbatch job ID here> <other options> -- cat /path/to/log```

This requires no ACLs or permission change and the entire log is streamed via SLURM. This doesn't solve the race condition issue (logs might have been touched by the user, most likely by accident), but it definitely solves the security gap and it less likely to open up one due to context differences between sites or by a future code change.

You could also submit your main job with ```srun``` rather than ```sbatch```, but the main job can, potentially, take a long time, and it's not the greatest to allocate a thread / process to get the output as it happens, quite rarely. Srun doesn't go to the background like sbatch does, it will instead connect stdin / stdout and stderr of the two ends. Streaming the long after the job is terminated makes it cleaner (when using srun) in my humble opinion. There also other reasons why you want to use sbatch for the actual job, as sbatch allows for a different set of features that could be useful. I also think it's a bit easier to handle errors.

ZeroMQ can be an option yes. SLURM uses [munge](https://github.com/dun/munge) for authentication, maybe you can do the same or do something similar. You can add a symmetric encryption key to the payload and use openssl for channel encryption if you wish. If you want it fancier, you can exchange public keys via munge instead and do authentication both ways again with openssl.

Thoughts?

**Comment by jluethi** [comment]
Thanks for the suggestions Enrico. 

The way we have Fractal server designed is that it can use a slurm runner to submit jobs, but it can also user other backends. Thus, choosing slurm specific features should be our last resort. For now, we just have a `process` runner (to allow users to run their own Fractal instances e.g. on their laptops and for testing and development. It directly submits jobs on a user's machine without slurm involvement) and the slurm runner. As far as possible, I'd like to avoid having code that is too custom for each runner.

Our main goal currently is to get Fractal running in a safe manner at the FMI without having to go into major refactors of the whole runner setup, if possible. We can't dedicate continuous resources on this issue at the level we've been doing for too long without other urgent priorities suffering. This is also why we would want to avoid or at least postpone a refactor towards a munge or ZeroMQ-based solution for the time being.

Are there security concerns or other blocking issues with the ACL approach that we're describing here and which, to us, is currently looking like a valid implementation to get Fractal running at FMI? These we'd obviously prioritize. Inputs on ideas towards improving the system, directions for a refactor to increase its stability or reduce the required maintenance or of course also appreciated, but we'd not prioritize those for the first implementation of a Fractal setup at FMI.

Also, to make sure we don't forget the context: We're talking about a contained folder where certain temporary files, virtual environments and logs are being saved. The main data (image data, measurements etc.) is being written by the correct user to a path in their share space that they have permission to without Fractal having to change any permissions to any of it.

**Comment by tcompa** [comment]
A side comment

> What about SLURM streaming at the beginning rather than ACLs, as a first simple implementation? Fractal can run any job as the user, so you can run:
> `srun -n 1 -c 1 --mem 1G --time '5:0' -dependency afterany:<sbatch job ID here> <other options> -- cat /path/to/log`
> This requires no ACLs or permission change and the entire log is streamed via SLURM. This doesn't solve the race condition issue (logs might have been touched by the user, most likely by accident), but it definitely solves the security gap and it less likely to open up one due to context differences between sites or by a future code change.

I am missing something here, or there is a misunderstanding.

The way jobs are submitted in Fractal is roughly via `sudo -u test01 sbatch /path/to/submit.sh`, and inside `/path/to/submit.sh` there is a `srun` command. We could try to skip `sbatch`, if that turns out to be useful, but I cannot see how this fixes the access-control issue. In the example with `-- cat /path/to/log`, the user `test01` needs permissions to write to `/path/to/log`, and `fractal` needs permissions to read that file. And also the inverse process is required (`fractal` writes, `test01` reads), likely in the same `/path/to` folder.

**Question:** How is this possible without ACL, without `/path/to` being 777 folder, and without using UNIX groups?

Unless I'm missing something trivial, this is exactly the problem we are trying to solve.

Briefly: because we want SLURM jobs to be run by specific user (this has always been a hard requirement for fractal), there will need to coexist files created by the `fractal` and `test01` users, in some folder. The question is how to restrict the access to this folder as much as possible. If there is no such issue, then we can just go with 777 folders and fully forget about ACL (which would be a simpler and more portable solution, agreed!).

**Comment by enricotagliavini** [comment]
Hello Tommaso, sorry for the delay in answering, I was busy with a time sensitive maintenance and managed to catch up only today talking with Gustavo and Joel, which then I understand met with you and already explained what we discussed. 

Thank you for looking into this.

**Comment by tcompa** [comment]
Closing by now, in favor of #473.
